src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHelper.cs

```diff
@@ -0,0 +1,313 @@
+        {
+            var solution = context.Solution;
+            Contract.ThrowIfNull(solution);
+
+            // TO-DO: We currently must execute code actions which add new documents on the server as commands,
+            // since there is no LSP support for adding documents yet. In the future, we should move these actions
+            // to execute on the client.
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1147293/
+
+            var textDiffService = solution.Services.GetService<IDocumentTextDifferencingService>();
+
+            using var _1 = ArrayBuilder<SumType<TextDocumentEdit, CreateFile, RenameFile, DeleteFile>>.GetInstance(out var textDocumentEdits);
+            using var _2 = PooledHashSet<DocumentId>.GetInstance(out var modifiedDocumentIds);
+
+            foreach (var option in operations)
+            {
+                // We only support making solution-updating operations in LSP.  And only ones that modify documents. 1st
+                // class code actions that do more than this are supposed to add the CodeAction.MakesNonDocumentChange
+                // in their Tags so we can filter them out before returning them to the client.
+                //
+                // However, we cannot enforce this as 3rd party fixers can still run.  So we filter their results to 
+                // only apply the portions of their work that updates documents, and nothing else.
+                if (option is not ApplyChangesOperation applyChangesOperation)
+                {
+                    context.TraceInformation($"Skipping code action operation for '{data.UniqueIdentifier}'.  It was a '{option.GetType().FullName}'");
+                    continue;
+                }
+
+                var changes = applyChangesOperation.ChangedSolution.GetChanges(solution);
+                var newSolution = await applyChangesOperation.ChangedSolution.WithMergedLinkedFileChangesAsync(solution, changes, cancellationToken: cancellationToken).ConfigureAwait(false);
+                changes = newSolution.GetChanges(solution);
+
+                var projectChanges = changes.GetProjectChanges();
+
+                // Don't apply changes in the presence of any non-document changes for now.  Note though that LSP does
+                // support additional functionality (like create/rename/delete file).  Once VS updates their LSP client
+                // impl to support this, we should add that support here.
+                //
+                // https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspaceEdit
+                //
+                // Tracked with: https://github.com/dotnet/roslyn/issues/65303
+                foreach (var projectChange in projectChanges)
```

Comment:
is this all a move from before?  or did you make changes?

Reply:
this is all from before